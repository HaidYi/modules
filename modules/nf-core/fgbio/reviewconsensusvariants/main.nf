process FGBIO_REVIEWCONSENSUSVARIANTS {
    tag "$meta.id"
    label 'process_low'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/87/87626ef674e2f19366ae6214575a114fe80ce598e796894820550731706a84be/data' :
        'community.wave.seqera.io/library/fgbio:2.4.0--913bad9d47ff8ddc' }"

    input:
    tuple val(meta), path(consensus_bam), path(consensus_bam_index), path(grouped_bam), path(grouped_bam_index), path(position_list)
    tuple val(meta2), path(fasta)
    tuple val(meta3), path(fasta_fai)
    tuple val(meta4), path(dict)

    output:
    tuple val(meta), path("*.grouped.bam")  , path("*.grouped.bai")  , emit: grouped_bam
    tuple val(meta), path("*.consensus.bam"), path("*.consensus.bai"), emit: consensus_bam
    tuple val(meta), path("*.txt")                                   , emit: summary
    path "versions.yml"                                              , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def mem_gb = 8
    if (!task.memory) {
        log.info '[fgbio FilterConsensusReads] Available memory not known - defaulting to 8GB. Specify process memory requirements to change this.'
    } else if (mem_gb > task.memory.giga) {
        if (task.memory.giga < 2) {
            mem_gb = 1
        } else {
            mem_gb = task.memory.giga - 1
        }
    }

    if ("${consensus_bam}" == "${prefix}.consensus.bam") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    if ("${consensus_bam}" == "${prefix}.grouped.bam")   error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    if ("${grouped_bam}"   == "${prefix}.consensus.bam") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
    if ("${grouped_bam}"   == "${prefix}.grouped.bam")   error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"

    """
    fgbio \\
        -Xmx${mem_gb}g \\
        --tmp-dir=. \\
        ReviewConsensusVariants \\
        --input ${position_list} \\
        --consensus-bam ${consensus_bam} \\
        --grouped-bam ${grouped_bam} \\
        --output ${prefix} \\
        --ref ${fasta} \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fgbio: \$(fgbio --version)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """

    touch ${prefix}.grouped.bam
    touch ${prefix}.grouped.bai
    touch ${prefix}.consensus.bam
    touch ${prefix}.consensus.bai
    touch ${prefix}.txt

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fgbio: \$(fgbio --version)
    END_VERSIONS
    """
}
