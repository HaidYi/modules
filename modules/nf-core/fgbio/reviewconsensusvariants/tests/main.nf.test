nextflow_process {

    name "Test Process FGBIO_REVIEWCONSENSUSVARIANTS"
    script "../main.nf"
    process "FGBIO_REVIEWCONSENSUSVARIANTS"

    tag "modules"
    tag "modules_nfcore"
    tag "fgbio"
    tag "fgbio/reviewconsensusvariants"

    test("homo_sapiens") {
        config "./nextflow.config"

        setup {
            run("SAMTOOLS_SORT", alias: "SAMTOOLS_SORT_GROUPED") {
                script "../../../samtools/sort/main.nf"
                process {
                    """
                    input[0] = [[ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/umi/test.paired_end.duplex_umi_grouped.bam', checkIfExists: true)
                                ]
                    input[1] = [[],[]]
                    """
                }
            }

            run("SAMTOOLS_SORT", alias: "SAMTOOLS_SORT_CONSENSUS") {
                script "../../../samtools/sort/main.nf"
                process {
                    """
                    input[0] = [[ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/umi/test.paired_end.duplex_umi_duplex_consensus.bam', checkIfExists: true)
                                ]
                    input[1] = [[],[]]
                    """
                }
            }

            run("BWA_MEM") {
                script "../../../bwa/mem/main.nf"
                process {
                    """
                    input[0] = [[ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/bam/umi/test.paired_end.duplex_umi_duplex_consensus.bam', checkIfExists: true)
                                ]
                    input[1] = [[],[]]
                    """
                }
            }

        }



        when {
            process {
                """

                input[0] =
                    SAMTOOLS_SORT_CONSENSUS.out.bam
                    .join(SAMTOOLS_SORT_CONSENSUS.out.csi)
                    .join(SAMTOOLS_SORT_GROUPED.out.bam)
                    .join(SAMTOOLS_SORT_GROUPED.out.csi)
                    .join(Channel.of([[ id: "test" ], file('/workspace/modules/test.intervals')]))

                input[1] = [
                    [ id:'homo_sapiens'], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true),
                ]
                input[2] = [
                    [ id:'homo_sapiens'], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true),
                ]
                input[3] = [
                    [ id:'homo_sapiens'], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("homo_sapiens - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    [],
                    [],
                    [],
                    [],
                    [],
                ]
                input[1] = [
                    [ id:'homo_sapiens'], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true),
                ]
                input[2] = [
                    [ id:'homo_sapiens'], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta.fai', checkIfExists: true),
                ]
                input[3] = [
                    [ id:'homo_sapiens'], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.dict', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
