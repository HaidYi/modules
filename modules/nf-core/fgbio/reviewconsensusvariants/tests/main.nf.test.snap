{
    "sarscov2 - bam - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "consensus_bam": [
                    
                ],
                "grouped_bam": [
                    
                ],
                "summary": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-13T11:13:27.743575602"
    },
    "homo_sapiens - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.grouped.bam:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test.grouped.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.consensus.bam:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test.consensus.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,43793ba25f3598f496659e8c434a009c"
                ],
                "consensus_bam": [
                    [
                        {
                            "id": "test"
                        },
                        "test.consensus.bam:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test.consensus.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "grouped_bam": [
                    [
                        {
                            "id": "test"
                        },
                        "test.grouped.bam:md5,d41d8cd98f00b204e9800998ecf8427e",
                        "test.grouped.bai:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "summary": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,43793ba25f3598f496659e8c434a009c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-13T11:18:59.058456817"
    },
    "homo_sapiens - bam": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "consensus_bam": [
                    
                ],
                "grouped_bam": [
                    
                ],
                "summary": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-13T11:15:56.53941388"
    },
    "homo_sapiens": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "consensus_bam": [
                    
                ],
                "grouped_bam": [
                    
                ],
                "summary": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-13T11:18:38.777634516"
    }
}